network/retrieve/retrieve_test.go

```diff
@@ -0,0 +1,403 @@
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/network/simulation"
+	"github.com/ethersphere/swarm/p2p/protocols"
+	"github.com/ethersphere/swarm/state"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+	"github.com/ethersphere/swarm/storage/mock"
+	"github.com/ethersphere/swarm/testutil"
+	"golang.org/x/crypto/sha3"
+)
+
+var (
+	loglevel           = flag.Int("loglevel", 5, "verbosity of logs")
+	bucketKeyFileStore = simulation.BucketKey("filestore")
+	bucketKeyNetstore  = simulation.BucketKey("netstore")
+
+	hash0 = sha3.Sum256([]byte{0})
+)
+
+func init() {
+	flag.Parse()
+
+	log.PrintOrigins(true)
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(os.Stderr, log.TerminalFormat(false))))
+}
+
+// TestChunkDelivery brings up two nodes, stores a few chunks on the first node, then tries to retrieve them through the second node
+func TestChunkDelivery(t *testing.T) {
+	chunkCount := 10
+	filesize := chunkCount * 4096
+
+	sim := simulation.NewBzzInProc(map[string]simulation.ServiceFunc{
+		"bzz-retrieve": newBzzRetrieveWithLocalstore,
+	})
+	defer sim.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	defer cancel()
+	_, err := sim.AddNode()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		nodeIDs := sim.UpNodeIDs()
+		log.Debug("uploader node", "enode", nodeIDs[0])
+
+		item := sim.MustNodeItem(nodeIDs[0], bucketKeyFileStore)
+
+		//put some data into just the first node
+		data := make([]byte, filesize)
+		if _, err := io.ReadFull(rand.Reader, data); err != nil {
+			t.Fatalf("reading from crypto/rand failed: %v", err.Error())
+		}
+		refs, err := getAllRefs(data)
+		if err != nil {
+			return nil
+		}
+		log.Trace("got all refs", "refs", refs)
+		_, wait, err := item.(*storage.FileStore).Store(context.Background(), bytes.NewReader(data), int64(filesize), false)
+		if err != nil {
+			return err
+		}
+		if err := wait(context.Background()); err != nil {
+			return err
+		}
+
+		id, err := sim.AddNode()
+		if err != nil {
+			return err
+		}
+		err = sim.Net.Connect(id, nodeIDs[0])
+		if err != nil {
+			return err
+		}
+		nodeIDs = sim.UpNodeIDs()
+		if len(nodeIDs) != 2 {
+			return fmt.Errorf("wrong number of nodes, expected %d got %d", 2, len(nodeIDs))
+		}
+
+		// allow the two nodes time to set up the protocols otherwise kademlias will be empty when retrieve requests happen
+		time.Sleep(50 * time.Millisecond)
+		log.Debug("fetching through node", "enode", nodeIDs[1])
+		ns := sim.MustNodeItem(nodeIDs[1], bucketKeyNetstore).(*storage.NetStore)
+		ctr := 0
+		for _, ch := range refs {
+			ctr++
+			_, err := ns.Get(context.Background(), chunk.ModeGetRequest, storage.NewRequest(ch))
+			if err != nil {
+				return err
+			}
+		}
+		if ctr != len(refs) {
```

Comment:
redundant if, this code would never be reached, as `ctr` will always be `len(refs)`, or we would have returned.